### PR TITLE
Fix JSON orient='table' issues with numeric column names

### LINKED ISSUE #19129: Raise ValueError for read_json and orient='table' With Numeric Column Names
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame([[1,2,3,4]], columns=[5,6,7,8])
df.to_json('test.json', orient='table')
pd.read_json('test.json', orient='table')

KeyError: '[5 6 7 8] not in index'
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 16d02621212cd7d83f89533265fc039df1d7be10
python: 3.6.3.final.0
python-bits: 64
OS: Darwin
OS-release: 17.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0.dev0+79.g16d026212
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: 0.26.1
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.8.0
xarray: 0.10.0
IPython: 6.2.1
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.1.1
openpyxl: 2.5.0b1
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.1.13
pymysql: 0.7.11.None
psycopg2: None
jinja2: 2.10
s3fs: 0.1.2
fastparquet: 0.1.3
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #22525: Round trip json serialization issues
#### Problem description
I would hope there would be a way for each of the serialization methods to store and retrieve enough metadata to reconstruct the original object.
Reading through #4889, it seems that most of the problems associated should have been resolved with 0.20 with the introduction of orient='table'.

When writing tests for this, it became apparent that several circumstances could produce issues.

tl;dr The schema as dumped and read is inconsistent, producing problematic behavior

#### Example 1: Empty DataFrame
```python
import pandas
from pandas.util.testing import assert_frame_equal
df = pandas.DataFrame()
assert_frame_equal(df, pandas.read_json(df.to_json(orient='table'), orient='table'))
AssertionError: DataFrame.index are different

DataFrame.index classes are not equivalent
[left]:  Float64Index([], dtype='float64')
[right]: Index([], dtype='object')

```
inspecting the to_json output we get
```python
df.to_json(orient='table')
'{"schema": {
    "fields":[{"name":"index","type":"string"}],
    "primaryKey":["index"],"pandas_version":"0.20.0"}, "data": []}'
```
According to http://pandas-docs.github.io/pandas-docs-travis/io.html#io-table-schema
the "type":"string" should indicate the pandas type of "object". The DataFrame.to_json is not respecting the type in memory and is converting the type from float64 to object.
#### Example 2: Int DataFrame
```python
import pandas
from pandas.util.testing import assert_frame_equal
df = pandas.DataFrame([1])
assert_frame_equal(df, pandas.read_json(df.to_json(orient='table'), orient='table'))

ValueError: Cannot convert non-finite values (NA or inf) to integer
```
I'm not sure exactly what is going on here.
If I drop the orient='table', it creates a valid DataFrame with equal values but still has the problem of still failing the assertion due to inferred index data types.
Inspecting the to_json output we get:
```python
df.to_json(orient='table')                                            
'{"schema": {
    "fields":[{"name":"index","type":"integer"},"name":0,"type":"integer"}],
    "primaryKey":"index"],"pandas_version":"0.20.0"}, "data": [{"index":0,"0":1}]}'
```
So now we correctly encode the type as integer but we can't decode it.
#### Example 3: Float DataFrame
```python
import pandas
from pandas.util.testing import assert_frame_equal
df = pandas.DataFrame([1.0])
assert_frame_equal(df, pandas.read_json(df.to_json(orient='table'), orient='table'))
AssertionError: DataFrame.iloc[:, 0] are different

DataFrame.iloc[:, 0] values are different (100.0 %)
[left]:  [1.0]
[right]: [nan]
```
Now we have a silent converting of the float value of 1.0 to nan. Probably related to the example 2 issue but in this case the index can handle nan as it is a float index instead of integer.
Inspecting the json output
```python
df.to_json(orient='table')                                            
'{"schema": {
    "fields:[{"name":"index","type":"integer"},"name":0,"type":"number"}],
    "primaryKey":"index"],"pandas_version":"0.20.0"}, "data": [{"index":0,"0":1.0}]}'                                                               
```

#### Code to replicate most cases that could fail

<details>

```python
from hypothesis import given, strategies as st
from hypothesis.extra.pandas import data_frames, column
from pandas.util.testing import assert_frame_equal
import pandas

# Generate dataframe columns
st_columns = st.builds(column, st.text(min_size=1), dtype=st.sampled_from([int, float, str, bool]))

# General a list of columns but ensure the column name is unique
st_lst_of_columns = st.lists(st_columns, unique_by=lambda x: x.name, min_size=1)

# Parse a list of random columns into the dataframe strategy
st_random_df = st_lst_of_columns.flatmap(data_frames)

@given(st_random_df)
def test_dataframe(df):
    assert_frame_equal(df, df)

@given(st_random_df)
def test_round_trip_json_dataframe(df):
    assert_frame_equal(df, pandas.read_json(df.to_json(orient='table'), orient='table'))
```

</details>

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS                                                        
------------------                
                                        
commit: None                                                              
python: 3.7.0.final.0                                                     
python-bits: 64                                                           
OS: Windows                                                               
OS-release: 10                                                            
machine: AMD64                                                            
processor: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel           
byteorder: little                                                         
LC_ALL: None                                                              
LANG: None                                                                
LOCALE: None.None                                                         
                                                                          
pandas: 0.23.4                                                            
pytest: 3.7.2                                                             
pip: 18.0                                                                 
setuptools: 40.2.0                                                        
Cython: None                                                              
numpy: 1.15.0                                                             
scipy: None                                                               
pyarrow: None                                                             
xarray: None                                                              
IPython: 6.5.0                                                            
sphinx: 1.7.7                                                             
patsy: None                                                               
dateutil: 2.7.3                                                           
pytz: 2018.5                                                              
blosc: None                                                               
bottleneck: None                                                          
tables: None                                                              
numexpr: None                                                             
feather: None                                                             
matplotlib: None                                                          
openpyxl: None                                                            
xlrd: None                                                                
xlwt: None                                                                
xlsxwriter: None                                                          
lxml: None                                                                
bs4: None                                                                 
html5lib: None                                                            
sqlalchemy: None                                                          
pymysql: None                                                             
psycopg2: None                                                            
jinja2: 2.10                                                              
s3fs: None                                                                
fastparquet: None                                                         
pandas_gbq: None                                                          
pandas_datareader: None                                                   

</details>

### PR DESCRIPTION
**Pull request by albertvillanova**
closes #19129
closes #22525
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/json/test_pandas.py [hunk 1]
```diff
@@ -1204,9 +1204,10 @@ def test_data_frame_size_after_to_json(self):
 
     @pytest.mark.parametrize('index', [None, [1, 2], [1., 2.], ['a', 'b'],
                                        ['1', '2'], ['1.', '2.']])
-    @pytest.mark.parametrize('columns', [['a', 'b'], ['1', '2'], ['1.', '2.']])
+    @pytest.mark.parametrize('columns', [None, [1, 2], [1., 2.], ['a', 'b'],
```
Comment (WillAyd) [inline pandas/tests/io/json/test_pandas.py]: So I don't know that we want to do this. Is it valid JSON in the table spec to have column names that are non-string?

Understood you have gotten this to round trip but if it violates the Table spec for JSON then I'd rather raise as commented previously

Comment (albertvillanova) [reply pandas/tests/io/json/test_pandas.py]: @WillAyd as I already commented in #19129, after having tried to start a discussion before doing the PR:

> I was suggesting not just "raising a more descriptive ValueError" (sic), but changing the implementation of the JSON serialization for orient='table'.

Could you please tell me where the JSON table spec claims that column names MUST be strings?

I am going to make a longer comment to further justify my PR.

**Comment by albertvillanova** [comment]
Hello, here my explanation to justify this PR.

As I already commented in #19129, IIUC, the only aim of orient='target' is to get round-trip JSON serialization/deserialization of pandas objects. Thus, the 'schema' was added following the "Table Schema" spec in https://frictionlessdata.io/specs/table-schema/.

This spec is ONLY for the 'schema', not for the table itself. And this spec does not prohibit the column names to be non-strings. Indeed, when serializing pandas DataFrames with non-string column names, their 'schema' is totally valid, and the issues arise just because of the JSON serialization of the table itself (the 'data').

In order to serialize the data itself, among all the possible orient values, it was chosen orient='records' and this is the cause of all the issues. Why? Because orient='records' serializes each DataFrame row as a JSON object (roughly equivalent to a Python dict), using the column names as the JSON object names (dict keys). The choice of using column names as dict keys (JSON object names) is not right for several reasons:
- non-string column names: 
DataFrame column names are not restricted to be strings. Indeed, non-string column names are the default if no explicit column names are passed to DataFrame instantiation; however, JSON spec imposes that JSON object names must be strings. 
- order of the columns:
DataFrame columns have a specific order; however, when representing each record as a Python dict, we lose the order of the columns.
- non-unique column names
a DataFrame can have non-unique column names; however a Python dict cannot have non-unique keys.

In order to address all theses issues, this PR proposes to use another representation of each DataFrame row: instead of using a Python dict, we should use a Python list. The Python list:
- can have non-unique items
- conserves the order of its items
- and its JSON equivalent is a JSON array, which does not impose any restriction on the types of its values

And in pandas, this can be easily made by choosing the orient='values' (instead of orient='records').

Finally, I would like to point out that there exists a Tabular Data Resource spec (https://frictionlessdata.io/specs/tabular-data-resource/) and it clearly establishes (in its section JSON Tabular Data):

> JSON Tabular Data MUST be an `array` where each item in the array MUST be:
> - EITHER: an array where each entry in the array is the value for that cell in the table
> - OR: an object where each key corresponds to the header for that row and the value corresponds to the cell value for that row for that header

This PR proposes to use the first option (instead of the second one) for all the previously mentioned reasons.

File: pandas/tests/io/json/test_pandas.py (hunk 1 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_pandas.py]: > Could you please tell me where the JSON table spec claims that column names MUST be strings?

Not specific to the table spec as much as just JSON itself. See the description of an object here:

https://json.org

**Comment by WillAyd** [comment]
> This PR proposes to use the first option (instead of the second one) for all the previously mentioned reasons.

OK I see your point here though so thanks for clarifying. Will review in more detail later and give you feedback just wanted to reach out in the interim

File: pandas/tests/io/json/test_pandas.py (hunk 1 above)
Comment (albertvillanova) [reply pandas/tests/io/json/test_pandas.py]: @WillAyd you are talking about different things:
- JSON spec imposes in a JSON object (composed of key-value pairs), its keys must be strings
- but we are talking about column names, not JSON object keys

And my question is: where the JSON table spec claims that COLUMN NAMES (not JSON object keys) must be strings?

**Review by jreback — CHANGES_REQUESTED** [review]
not averse to doing this, but this completely breaks backward compatibility. So. update the version on the table. Also I think need to save some samples using the original format and add some tests to be able to read them (dispatch on the version).

File: pandas/tests/io/json/test_pandas.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/tests/io/json/test_pandas.py]: @pwalsh can you comment on this? Is the `name` field in a Field Descriptor expected to be a string?

**Review by TomAugspurger — COMMENTED** [review]
We can't just change the output like this without a deprecation.

In retrospect, it seems like we overloaded the `orient` keyword to handle two things

1. whether or not we include a JSON table schema
2. the orient of the data within the output

What do people think about leaving `orient` just for the data key, and adding a new keyword `schema=True/False` to control whether we include the `schema` key at the outer level? I haven't really thought through the deprecation cycle.

**Comment by WillAyd** [comment]
> What do people think about leaving `orient` just for the data key, and adding a new keyword `schema=True/False` to control whether we include the `schema` key at the outer level? I haven't really thought through the deprecation cycle.

What do you see as the advantage of doing this? I would actually be against this as we already have quite a few JSON formats, so adding another that doesn't follow the table spec doesn't add a lot of value I think.

**Comment by TomAugspurger** [comment]
I’ll need to read the spec again, but it may not make any claims for how the data should be represented.

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
can you merge master

**Comment by WillAyd** [comment]
Not a bad idea but this change is definitely breaking and comes with quite a few implications. Given it's gone somewhat stale over the past month and a half going to close out, but feel free to ping back @albertvillanova if it's something you want to drive forward
